### PR TITLE
[Do not merge]CORTX-29587 : Design for FKVV node format to hold embedded records with indirect ad…

### PR DESCRIPTION
**Pull request by RadhaGulhane13**
…dressing

Signed-off-by: Radha Gulhane <radha.gulhane@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: btree/btree.c [hunk 1]
```diff
@@ -3346,6 +3346,58 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ *  |  |      |      |  |      |      |  |     |     |                   |        |        |        |
+ *  |  |  k0  |P_V0  |  | K1   | P_V1 |  |  K2 | P_V2| ---->             |  V2    |  V1    |   V0   |
+ *  | ^|      |      |  |      |      |  |     |     |             <---- |        |        |        |
+ *  | ||      |      |  |      |      |  |     |     |                   |        |        |        |
+ *  | ||      |      |  |      |      |  |     |     |                   |        |        |        |
+ *  +-++------+--+---+--+------+--+---+--+-----+--+--+-------------------+----^---+-----^--+----^---+
+ *    |          |                |               |                           |         |       |
+ *    |          |                |               +---------------------------+         |       |
+ *    |          |                +---------------+-------------------------------------+       |
+ *    |          +------------------------------------------------------------------------------+
```
Comment (kanchan-chaudhari) [inline btree/btree.c:3379]: 80 char limit, can you remove extra spaces in between columns?

Comment (kanchan-chaudhari) [reply btree/btree.c:3379]: Also what does first column in diagram indicate?? Node Header??? Please correct the straight line in first column.

File: btree/btree.c [hunk 2]
```diff
@@ -3346,6 +3346,58 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * reduce the movement of records and transaction capturing, we design FKVV node
+ * format which provided embedded keys and values with indirect addressing.
+ *
+ * We will use following directory structure for key :
+ * struct dir_key {
+ *     uint32_t  validate;
+ *     uint32_t  val_size;
+ *     ---------- constant space for key ------------
+ *     void     *p_val;
+ *     };
```
Comment (kanchan-chaudhari) [inline btree/btree.c:3366]: Align `};` to left.

Comment (kanchan-chaudhari) [inline btree/btree.c:3366]: Align `};` to left.

File: btree/btree.c (hunk 1 above)
Comment (RadhaGulhane13) [reply btree/btree.c]: Updated diagram

**Review by kanchan-chaudhari — APPROVED** [review]

File: btree/btree.c [hunk 3]
```diff
@@ -3346,6 +3346,113 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
...
+ * Node will also have offset directory. Initially, this directory will be
+ * situated at the middle of node but if we encounter the possibilty for
+ * overlapping of keys/values with offset directory, we will move the
+ * entire direcory to left or right.
+ *
+ * All directory entries are sorted according to the keys that they point to.
+ * i.e *orec1->kptr < *orec1->kptr < *orec1->kptr
+ *
+ * When inserting a new record, we will find the space between any two records
+ * such that it is able to accomodate new record. Bitmaping can be used to keep
```
Comment (shashank-parulekar) [inline btree/btree.c:3396]: We do not plan to use bitmap for free records. Instead we will track the total valid records and total max records. All the records between total max records and total valid records contain entries which track the free slots.

**Review by shashank-parulekar — CHANGES_REQUESTED** [review]

**Review by shashank-parulekar — APPROVED** [review]

**Pull request closed without merging**